haskell/README.md

```diff
@@ -63,51 +109,63 @@ with lenses we can do this
 lexer' = lexer & input %~(take 2) & position +~1 & readPosition +~1 & ch .~ 'b'
 ```
 
-This seems trivial here but when if we have a record with many fields it becomes very verbose and hard to read particularly with nested records.
-
-## [Data.ByteString.Char8](https://hackage.haskell.org/package/bytestring-0.11.4.0/docs/Data-ByteString-Char8.html)
+This seems trivial here but when if we have a record with many fields it
+becomes very verbose and hard to read particularly with nested records.
 
-Haskell `String` type is actually a list of `char`.
-This is not very performant, e.g., `length` and indexing `(!!)` operations are O(N). Also, indexing is unsafe (can fail) with crafting or importing safe operations. With `ByteString` we have O(1) operations and built in safe indexing that return `Maybe` types (Option in Rust).
+### [Data.ByteString.Char8](https://hackage.haskell.org/package/bytestring-0.11.4.0/docs/Data-ByteString-Char8.html)
 
-## [Data.Maybe](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Maybe.html)
-
-Used to unwrap `Maybe` type (with default in our case).
-
-## [Data.Char](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Char.html)
-
-Used  for built in `isSpace` , `isDigit`, `isLetter`
+Haskell `String` type is actually a linked list of `char`. This is not very
+performant, e.g., `length` and indexing `(!!)` operations are O(N). Also,
+indexing is unsafe (can fail) with crafting or importing safe operations. With
+`ByteString` we have O(1) operations and built in safe indexing that return
+`Maybe` types (Option in Rust).
 
 ## [Control.Monad.Trans.State](https://hackage.haskell.org/package/transformers-0.6.1.0/docs/Control-Monad-Trans-State.html)
 
-since we are working with immutable variables, we have to keep passing the state around. You can see this in the `OCaml` version where `lexer` is being parsed around recursively and returned in a tuple?.
+Since we work with pure function (that have no side effects), we have to keep
+passing the state around. You can see this in the `Lexer.Basic` version where
+the lexer (a simple `String` in that case) is being parsed around recursively
+and returned in a tuple `nextToken :: Lexer -> (Token, Lexer)`. In languages
+like Rust we can implement a function for a structure to mutate it `pub fn
+next_token(&mut self) -> Token`.
 
-The `state` monad allows us to obfuscate the passing around of the state and perform updates on the state within a function without the need for new variable names.
+The `state` monad allows us to obfuscate the passing around of the state and
+perform updates on the state within a function without the need for new
+variable names.
```

Comment:
I would still add that it allows us to simulate mutability without actually having mutability. "Passing around of the state" implies that we have a state that we pass around, when in "reality" (ignoring compiler optimizations) we create a new state for every function call.

Reply:
I see your point. I will try to find a better explanation for this one.